rust/signed_doc/src/metadata/mod.rs

```diff
@@ -231,33 +232,77 @@ impl Display for Metadata {
+            !self.extra().collabs().is_empty(),
+            self.extra().parameters().is_some(),
+        ]
+        .iter()
+        .filter(|v| **v)
+        .count();
 
-    fn try_from(meta: &Metadata) -> Result<Self, Self::Error> {
-        let mut builder = coset::HeaderBuilder::new()
-            .content_format(CoapContentFormat::from(meta.content_type()?));
-
-        if let Some(content_encoding) = meta.content_encoding() {
-            builder = builder.text_value(
-                CONTENT_ENCODING_KEY.to_string(),
-                format!("{content_encoding}").into(),
-            );
+        e.map(
+            number_of_fields
+                .try_into()
+                .map_err(minicbor::encode::Error::message)?,
+        )?;
+        if let Some(content_type) = &self.0.content_type {
+            e.encode(3)?.encode(content_type)?;
+        }
+        if let Some(content_encoding) = &self.0.content_encoding {
+            e.str(CONTENT_ENCODING_KEY)?.encode(content_encoding)?;
+        }
+        if let Some(doc_type) = &self.0.doc_type {
+            e.str(TYPE_KEY)?.encode(doc_type)?;
+        }
+        if let Some(id) = &self.0.id {
+            e.str(ID_KEY)?
+                .encode_with(id, &mut catalyst_types::uuid::CborContext::Tagged)?;
+        }
+        if let Some(ver) = &self.0.ver {
+            e.str(VER_KEY)?
+                .encode_with(ver, &mut catalyst_types::uuid::CborContext::Tagged)?;
+        }
+        if let Some(doc_ref) = &self.extra().doc_ref() {
+            e.str(REF_KEY)?.encode(doc_ref)?;
+        }
+        if let Some(template) = &self.extra().template() {
+            e.str(TEMPLATE_KEY)?.encode(template)?;
+        }
+        if let Some(reply) = &self.extra().reply() {
+            e.str(REPLY_KEY)?.encode(reply)?;
+        }
+        if let Some(section) = self.extra().section() {
+            e.str(SECTION_KEY)?.encode(section)?;
         }
 
-        builder = builder
-            .text_value(TYPE_KEY.to_string(), meta.doc_type()?.to_value())
-            .text_value(
-                ID_KEY.to_string(),
-                Value::try_from(CborUuidV7(meta.doc_id()?))?,
-            )
-            .text_value(
-                VER_KEY.to_string(),
-                Value::try_from(CborUuidV7(meta.doc_ver()?))?,
-            );
-
-        builder = meta.0.extra.fill_cose_header_fields(builder)?;
-
-        Ok(builder.build())
+        if !self.extra().collabs().is_empty() {
+            e.str(COLLABS_KEY)?.array(
+                self.extra()
+                    .collabs()
+                    .len()
+                    .try_into()
+                    .map_err(minicbor::encode::Error::message)?,
+            )?;
+            for collab in self.extra().collabs() {
+                e.str(collab)?;
+            }
+        }
+        if let Some(parameters) = self.extra().parameters() {
+            e.str(PARAMETERS_KEY)?.encode(parameters)?;
+        }
+        Ok(())
```

Comment:
There should be a CBOR helper which can be used when decoding a map to determine that the map keys are deterministically ordered.
We should have a unit test which encodes all possible protected headers using the above code to a map, and then decodes it, and uses this helper to ensure all the keys are in the correct deterministic order.

This will ensure that the order keys are added to the map is always correct, and prevent edits from breaking the correct sequence.